cirq/linalg/decompositions.py

```diff
@@ -1045,3 +1046,34 @@ def _gamma(u: np.ndarray) -> np.ndarray:
         u @ yy @ u.T @ yy, where yy = Y ⊗ Y
     """
     return u @ YY @ u.T @ YY
+
+
+def extract_right_diag(u: np.ndarray, atol=1e-15):
+    """Extract a diagonal unitary from a 3-CNOT two-qubit unitary.
+
+    Returns a 2-CNOT unitary D that is diagonal, so that U @ D needs only
+    two CNOT gates in case the original unitary is a 3-CNOT unitary.
+
+    See Proposition V.2 in Minimal Universal Two-Qubit CNOT-based Circuits.
+    https://arxiv.org/abs/quant-ph/0308033
+
+    Args:
+        u: three-CNOT two-qubit unitary
+        atol: the absolute tolerance for avoiding division by zero
+    Returns:
+        diagonal extracted from U
+    """
+    t = _gamma(transformations.to_special(u).T).T.diagonal()
+    k = np.real(t[0] + t[3] - t[1] - t[2])
+    if np.abs(k) < atol:
+        # in the end we have to pick a psi that makes sure that
+        # exp(-i*psi) (t[0]+t[3]) + exp(i*psi) (t[1]+t[2]) is real
+        # both pi/2 or 3pi/2 can work
+        psi = np.pi / 2
+    else:
+        psi = np.arctan(np.imag(np.sum(t)) / k)
+
+    cnot = block_diag(np.eye(2), np.array([[0, 1], [1, 0]]))
+    rz = np.diag([1, np.exp(1j * psi)])
+
+    return cnot @ combinators.kron(np.eye(2), rz) @ cnot
```

Comment:
CNOT is a permutation gate so matrix multiplications are unnecessary: it's action on a diagonal matrix is just a swap of the last two elements. Therefore, we can remove `cnot` and `rz` and can instead say

```python
f = np.exp(1j * psi)
return np.diag([1, f, f, 1])
```